4단계 - 로또(2등)

popotec opened the pull request:
안녕하세요. 4단계 리뷰 요청드립니다.

3단계 처음 제가 설계했던 코드를 생각하면 중구난방이었단 생각이 듭니다.
스스로 하고도 무언가 찜찜한 부분이 많았었는데... 리뷰어님이 정말 디테일하게(그렇게 짠 것이 죄송했습니다ㅠㅠ)
코멘트 해주시고, 한단계 더 나아갈 수 있는 부분들 가이드 해주셔서 3단계를 만족하며 끝낼수 있었습니다.

특히 이론으로 공부하면서 실제로 어떻게 적용할수 있을지 몰랐던 것들을 적용해본 게 신기하기도 하고.. 뿌듯했습니다.
- static 블록을 사용하여 캐싱 처리
- ```(Collections.shuffle)에 의존하는 코드를 의존성 주입을 통해 해결``` => 전략패턴을 적용하여 관심사 분리
(특히 기억에 남아 기록하고 싶어 열거해보았습니다..ㅎㅎ)

주말에도 신경써주셔서 감사합니다!

popotec committed:
refactor: LottoNumbersMaker, LottoTicketIssuer의 다른 객체에 대한 의존성 완화

1. LottoNumbersMaker에 List<LottoNumber>를 arrange하는 것을 전략패턴을 적용
- ArrangeManipulator라는 추상 전략 인터페이스 정의
- ArrangeManipulator의 구체 클래스인 LottoArrangeManipulator 정의
- LottoArrangeManipulator에서 실제로 List의 shuffle을 담당

2. LottoNumbersMaker를 인스턴스화 해서 사용
- 기존엔 인스턴스 변수에 의존하는 로직이 없어서 모든 메서드를 static으로 정의하여 사용하였었는데, 전략패턴을 적용하게 되면서, 주입받은 전략패턴을 사용하기 위해 메서드들을 인스턴스 메서드로 변경

3. LottoTicketIssuer의 LottoNumbersMaker 의존성 완화
- LottoTicketIssuer도 LottoNumbersMaker와 유사하게 메서드들을 모두 static으로 정의하여 사용하고 있었는데, LottoNumbersMaker을 객체화해서 사용하게 변경하며 자연스럽게 LottoTicketIssuer도 LottoNumbersMaker를 주입받아 사용하게 됨

==> 기존에는 LottoNumbersMaker와 LottoTicketIssuer가 유틸성 성격을 띄면서, 변화를 고려하지 못하고 있었는데, 해당 클래스에서 분리할 수 있는 관심사를 주입받도록 변경하면서 보다 변경에 유연해지고 확장성이 증가함

popotec committed:
feat(LottoMatchRank): 보너스 번호 매칭 유무에 따른 등급 추가

popotec committed:
test(LottoMatchRank): 기존 테스트에 보너스 매칭 케이스 추가

popotec committed:
refactor(LottoNumbersMaker, LottoTicket): 로또 번호들을 오름차순으로 정렬하는 것을 LottoTicket 생성자로 이동

- LottoNumbersMaker를 통하지 않고, LottoTicket의 생성자에 List<LottoNumber>를 전달하는 경우에는 Collections.sort가 실행되지 않았음
- 티켓 안의 번호를 순서대로 나타내고 싶은 것은 티켓의 요건이므로 티켓으로 역할을 옮겨도 괜찮다고 판단

popotec committed:
feat(BonusNumber): 보너스 번호 생성

popotec committed:
feat(BonusNumber): 보너스 번호 매칭 기능

popotec committed:
feat(LottoPrizeAnswer,BonusNumber): 보너스 번호를 포함한 당첨번호 생성

popotec committed:
feat(LottoPrizeAnswer): 당첨정답과 티켓 비교 기능

popotec committed:
refactor(LottoPrizeAnswer..): 당첨정답 비교 기능 리팩토링

- LottoWinningNumbers가 담당하던 정답과의 비교를 LottoPrizeAnswer로 역할 변경
- LottoPrizeAnswer는 LottoWinningNumbers와 BonusNumber와의 비교를 함께 하여 결과 제출

popotec committed:
refactor(LottoMatchRank): 2등인 경우와 toString 문구 나누기

popotec committed:
chore: 불필요 주석처리 구문 삭제

popotec committed:
feat: 당첨번호와 보너스 볼 숫자 중복되는 경우 예외 처리

popotec committed:
feat: UI에서 보너스 번호 입력받기 추가

popotec committed:
refactor: LottoWinningNumbers -> WinningNumbers 로 클래스명 변경

popotec committed:
refactor: 패키지내 접근만 필요한 메서드들 접근제어자 변경 public -> protected

popotec committed:
docs(README): 보너스 번호 기능 정의 추가

minsungryu commented on src/main/java/lottoservice/lottonumber/LottoNumbersMaker.java:
```diff
@@ -21,30 +22,35 @@
-	public static List<LottoNumber> makeLottoNumbers() {
-		Collections.shuffle(lottoNumberCandidates);
+	public List<LottoNumber> makeLottoNumbers() {
+		arrangeManipulator.shuffleElements(lottoNumberCandidates);
```
주석의 내용과 코드의 내용이 일치하지 않는데요 주석은 꼭 필요한 경우에 달아주시는게 좋을 것 같습니다 :)

minsungryu replied:
전에 리뷰드린 내용을 잘 반영해주셨네요 👍 
매번 피드백을 잘 반영해주셔서 감사합니다.

관련해서 한 가지 더 의견드릴게 있는데요,
저는 `LottoNumbersMaker` 입장에서 `makeLottoNumbers()`를 정의할 때 중요한건
숫자를 섞는 행위보다는 **숫자를 얻는 행위**일 것이라고 생각했습니다.

- 어떻게 숫자를 얻을 것인가?
- 어떤 숫자를 얻을 것인가?
- 숫자를 얼마나 많이 얻을 것인가?

등등 고민해볼 부분이 많은 것 같습니다.
정답은 없겠지만.. 객체를 설계할 때 이런 부분들을 고민해보시면 도움이 되지 않을까 싶습니다 :)

popotec replied:
넵 민성님이 조언 주신 부분에 따라 생각해보고 저도 **숫자를 얻는 행위** 라는 것에 공감했습니다.
숫자를 어떻게 중복되지 않게 6개를 추출하느냐는 Maker의 관심사에서 분리하여 추상화 시키는것이 훨씬 유연해진다 느꼈습니다.
```java
public List<LottoNumber> makeLottoNumbers() {
		arrangeManipulator.shuffleElements(lottoNumberCandidates);
		List<LottoNumber> lottoNumbers = new ArrayList<>(lottoNumberCandidates.subList(0,
			SIZE_OF_LOTTERY_NUMBERS));  /*List의 subList는 deepCopy가 아니므로 새로 인스턴스를 생성하여 전달 */
		return lottoNumbers;
	}
```
조금 더 생각해보니 위 코드에서 0번째부터 6개를 추출하는 부분까지도 arrangeManipulator에서 처리하는 것도 괜찮지 않을까하여 변경해보았습니다.

```java
public interface ArrangeManipulator {
...
	<T> List<T> getRandomElements(List<T> elements, int sizeOfPick);
}

public class LottoArrangeManipulator implements ArrangeManipulator {
...
@Override
	public <LottoNumber> List<LottoNumber> getRandomElements(List<LottoNumber> elements, int sizeOfPick) {
		shuffleElements(elements);
		return new ArrayList<>(elements.subList(0,	sizeOfPick));
	}
}

public class LottoNumbersMaker {
...
public List<LottoNumber> makeLottoNumbers() {
		return arrangeManipulator.getRandomElements(lottoNumberCandidates,SIZE_OF_LOTTERY_NUMBERS);
	}
...
}
``` 
이렇게 하면 극단적인 예로,
순차적인 인덱스로 6개를 추출하지 않고 2의 배수로 6개를 추출하는 등으로 방법을 변경하는 경우에도  LottoNumberMaker는 변경하지 않고 유연하게 프로그램을 수정할 수 있을 것 같습니다:)

minsungryu replied:
좋은 접근인 것 같습니다. 관련해서는 피드백을 추가로 남겼으니 확인해보시면 좋을 것 같아요 :)

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicket.java:
```diff
@@ -23,9 +24,14 @@ private LottoTicket() {
 
 	public LottoTicket(List<LottoNumber> lottoNumber) {
 		validateLottoNumberGroupRule(lottoNumber);
+		sortNumbersInAscOrder(lottoNumber);
```
생성자에서는 정렬을 하기보단 정렬이 되어있는지 검증하면 어떨지 생각이 들었습니다.
정렬되지 않은 Collection(Set, List)로 초기화하고 싶은 경우는 별도의 정적 팩토리 메서드를 통해서 기능을 제공할 수 있을 것 같습니다.

popotec replied:
로또 번호 정렬을 처음엔 LottoNumberMaker에서 했었는데, 생각하다보니 로또번호를 오름차순으로 정렬하고싶은 것은 LottoNumberMaker보다  LottoTicket의 니즈라고 생각하였습니다. (사용자에게 발매된 로또 티켓 안 번호를 오름차순으로 정렬하여 제공하기 위해)
LottoNumberMaker 입장에서는 List<LottoNumber>를 오름차순으로 정렬하는 것은 중요하지 않고 중복되지 않은 6개의 로또 숫자를 생성하는 것에만 관심을 가지면 된다 생각하였습니다.

그래서 LottoTicket을 생성할 때 정렬하도록 로직을 옮겼는데, 생성자가 과한 작업을 하게 만든 것일까요~?

minsungryu replied:
저는 오히려 `LottoNumberMaker`가 숫자를 만들고 정렬하는데 관심을 가져야한다고 생각했습니다.
말 그대로 생성을 해주는(Maker) 책임을 가지고 있으니까요.

`LottoTicket`은 단순히 "6개의 중복되지 않은 숫자를 오름차순으로 가지고 있다"는 명제를 잘 지키는 클래스면 충분하다고 생각했습니다. 그래서 외부에서 생성자나 팩토리 메서드를 통해 초기화를 시도할 경우 잘못된 입력값(중복, 순서 틀림, 숫자가 적음, 숫자가 많음)을 받는다면 생성자에서 유효성 검사를 통해 인스턴스화를 방지하는 정도로만 책임 범위를 생각했었어요.

---

(사람마다 견해가 갈리는 부분이라서 일단은 제 개인적인 생각을 말씀드립니다.)
생성자는 가급적이면 로직이 들어가지 않는게 좋습니다.
유효성 검사는 불완전한 객체가 생성되지 않기 위해 들어가는 것이니 어쩔 수 없이 타협을 하는 부분이라고 생각을 해요.
물론 유효성 검사가 철저한 경우 오히려 테스트 데이터를 준비하기 힘들어지는 케이스도 있습니다.

minsungryu commented on src/main/java/lottoservice/lottoticket/LottoTicket.java:
```diff
@@ -48,11 +54,14 @@ public boolean hasLottoNumber(LottoNumber lottoNumber) {
-		return lottoNumbers.stream()
-			.map(lottoNumber -> lottoNumber.toString())
-			.collect(Collectors.joining(", ", "[", "]"));
+	private boolean hasAllSameNumbers(LottoTicket that) {
```
equals를 구현하기 위해 노력해주셨네요 :)
다만 equals를 구현하기 위해서는 상당히 많은 부분에 대해 고민해봐야 합니다(이펙티브 자바 관련 포스팅 참고: https://jinioh88.tistory.com/36)

다행히도 IDE나 Lombok같은 플러그인이 이런 기능들을 제공하는데요 이를 활용해보면 어떨까요?
(LottoTicket이 가진 속성인 List의 equals로 위임할거에요)

minsungryu commented on src/main/java/lottoservice/matcher/BonusNumber.java:
```diff
@@ -0,0 +1,62 @@
+import lottoservice.lottonumber.LottoNumber;
+import lottoservice.lottoticket.LottoTicket;
+
+public class BonusNumber {
```
모든 원시값을 포장하라는 규칙을 잘 지켜주신 것 같습니다 👍 

한 가지 더 고민해보면 좋을 것 같은데요
제가 볼 땐 사실상 BonusNumber는 LottoNumber와 같은 클래스라고 보입니다.
`WinningNumbers`에 속성을 추가해서 문제를 해결할 수도 있지 않을까요?
```java
public class WinningNumbers {

    private final Set<LottoNumber> winningNumbers;
    private final LottoNumber bonusNumber;
    // ...
}
```
어떻게 생각하시는지 궁금합니다.

popotec replied:
저도 WinningNumbers에 BonusNumber를 추가하면 변경을 최소화하면서 보너스번호 매칭도 할 수 있을것 같아서 고려해보았었습니다.
(이렇게 할 경우 지금처럼 별도 클래스를 따로 만들어서 구조를 변경하는 일 없이 간단하게 추가 요구사항을 넣을 수 있을 것 같았습니다)
그런데 일급컬렉션 정의에 대해 고민해보니 [참고](https://developerfarm.wordpress.com/2012/02/01/object_calisthenics_/) 컬렉션 외 다른 멤버변수를 추가하면 안될것 같아서 LottoPrizeAnswer 라는 클래스를 새로 넣게 되었습니다.
=> 이 정의를 제가 해석하기로는 일급컬렉션으로 포장되면서 컬렉션이 하나의 identity를 갖고 불변성을 가지는 것이 특성이여서 다른 멤버변수와 조합될 수 없다고 생각하였습니다.

BonusNumber의 경우는 원시타입 int를 포장하는 것도 아니고 LottoNumber를 한번더 포장하는 것이여서 찜찜하긴 했습니다.
LottoNumber 타입을 사용하고 변수명을 bonusNumber라 하는 것도 생각해 보았는데요,
포비님이 변수명으로 특성을 나타내려하면 안 되다고 하셨던게 고민되었습니다.
BonusNumber가 LottoNumber와 기능적으로는 유사하지만 논리적인 정의로는 일반적인 로또 번호의 특성과는 차이가 있어서 , 변수명으로 나타내는 것이 적절치 않다 생각하였습니다. (특히나 matchTicket()와 같은 기능도  LottoNumber와 구별되는 특성을 부여한다 생각하였습니다.)

그래서 BonusNumber->LottoNumber 상속도 생각해 보았었습니다. 이 경우에 BonusNumber가 LottoNumber의 모든 것을 대체할 수 있고 확장된 클래스인가? 를 따져보았습니다. BonusNumber가 LottoNumber의 값 특성과 제약조건을 모두 가지고 있어서 대체 가능하고 상속 관계가 성립한다 생각하였습니다. 그런데 [prefer composition over inheritance](https://en.wikipedia.org/wiki/Composition_over_inheritance) 를 생각해보았을때 상속으로 결합을 강하게 가져가는 것보다 composition으로 유연성을 높이는게 낮지 않을까 하여 지금과 같은 설계가 되었습니다. ^^;

minsungryu replied:
네 많은 고민이 느껴지는 코드였습니다.
그에 반해 저의 피드백이 부족했던 것 같습니다.
제가 드린 피드백은 원칙에서 살짝 벗어난 현실과 타협했던 피드백 같네요^^;

현재 과정에서는 객체지향 생활체조 원칙을 지키기 위해 조금 더 고민을 해봐야할텐데요
이런 경우 일급 컬렉션을 선언하고, 그 일급 컬렉션을 감싸는 새로운 데이터 클래스를 만들어 볼 수 있을 것 같습니다.

```java
public class LottoNumbers {

    private final Set<LottoNumber> lottoNumbers;
}

public class WinningNumbers {

    private final LottoNumbers winningNumbers;
    private final LottoNumber bonusNumber;
}
```

minsungryu commented on src/main/java/lottoservice/matcher/LottoMatchRank.java:
```diff
@@ -7,7 +7,17 @@
  * 당첨번호와 비교하여 발생할 수 있는 결과 케이스를 enum으로 정의
  */
 public enum LottoMatchRank {
+	// FIRST(6, 2_000_000_000),
```
더 이상 사용하지 않는 주석은 바로바로 지워주셔도 좋을 것 같습니다 :)

minsungryu commented on src/main/java/lottoservice/matcher/LottoMatchRank.java:
```diff
@@ -32,10 +42,28 @@ public int getWinningMoney() {
 		return foundRank.orElseThrow(() -> new IllegalStateException(ERROR_MESSAGE_UNEXPECTED_COUNTOFMATCH));
 	}
+
+	private static boolean isFind(LottoMatchRank rank, int countOfMatch, boolean matchBonus) {
```
static 없이 인스턴스 메서드로 변경한다면 인자 수도 줄고 로직도 더 간단해질 수 있을 것 같아요!
또한 그 자체로 외부에 공개 가능한 public 메서드로 활용될 수 있지 않을까 생각도 듭니다.

minsungryu commented on src/test/java/lottoservice/matcher/WinningNumbersTest.java:
```diff
@@ -0,0 +1,88 @@
+	LottoNumbersMaker lottoNumbersMaker = new LottoNumbersMaker(new LottoArrangeManipulator());
+
+	@Test
+	public void makeLottoWinningNumbers_당첨번호_생성() {
```
junit5를 사용중인데 `@DisplayName`을 활용해서 테스트 이름을 좀 더 편하게 만들어보면 어떨까요?
아래 `@ParameterizedTest`의 경우 치환 문자열을 통해서 테스트 이름을 좀 더 풍부하게 만들어줄 수도 있습니다.

minsungryu commented on src/test/java/lottoservice/matcher/LottoPrizeAnswerTest.java:
```diff
@@ -0,0 +1,126 @@
+		for (int i = 0; i < 20; i++) {
+			tickets.add(new LottoTicket(lottoNumbersMaker.makeLottoNumbers(Arrays.asList(1, 6, 11, 25, 32, 44))));
+		}
+		tickets.add(new LottoTicket(lottoNumbersMaker.makeLottoNumbers(Arrays.asList(1, 4, 12, 26, 30, 40))));
```
`new LottoTicket(Arrays.asList(1, 4, 12, 26, 30, 40))` 대신 `lottoNumbersMaker`를 통해 초기화하신 이유가 있으신걸까요?

minsungryu replied:
이 테스트는 테스트를 위한 준비과정 코드가 긴 편인데요
테스트 픽스처 관련해서는 아래 포스트를 읽어보시면 도움이 될 것 같습니다.
https://jojoldu.tistory.com/611?category=1011740

popotec replied:
좋은 글 감사합니다.
준비과정이 길고 복잡한것에 어떻게 개선하는 것이 좋은가 고민하였었는데, 좋은 레퍼런스가 되었습니다.

minsungryu commented on src/test/java/lottoservice/lottoticket/LottoTicketTest.java:
```diff
@@ -16,12 +19,23 @@
+		List<LottoNumber> lottoNumbers = Arrays.asList(1,34,10,2,30,6)
+			.stream().map(it -> LottoNumber.valueOf(it)).collect(Collectors.toList());
+		LottoTicket lottoTicket = new LottoTicket(lottoNumbers);
+		for(int i=1; i<lottoTicket.getNumOfNumbersInGroup(); i++){
+			assertThat(lottoTicket.getLottoNumber(i).compareTo(lottoTicket.getLottoNumber(i-1))).isEqualTo(1);
+		}
```
테스트 안에 복잡해보이는 코드들이 보입니다.
테스트 가독성을 높일 수 있는 방법이 있을까요?
지금 생각나는 건 앞서 리뷰드린 것 처럼 정렬되지 않은 숫자로는 초기화되지 않도록 설계를 바꿔보는 방법도 있을 것 같습니다.
그게 아니라면 equals 메서드로 비교해보는 방법도 있겠네요.

minsungryu reviewed (changes requested):
지희님 로또 4단계 미션 구현하시느라 고생많으셨습니다!

매번 피드백 잘 반영해주셔서 고맙습니다. 저도 힘이 나네요!
물론 제 피드백이 정답은 아니기 때문에 무작정 따르실 필요는 없고
한 번 더 고민하는 계기가 되었으면 좋겠습니다.

이번에도 몇 가지 개선하면 좋을 것 같은 부분에 대해 코멘트 남겼습니다.
확인해주시면 좋을 것 같아요 :)

minsungryu commented:
> (그렇게 짠 것이 죄송했습니다ㅠㅠ)

그렇게 생각하지 않으셔도 돼요!!
저도 똑같은 한 명의 개발자일 뿐이고 어떻게 하면 더 좋을지 같이 고민하며 의견을 드리는 것 뿐입니다 😄

popotec committed:
chore: 불필요 주석 삭제

popotec committed:
refactor: equals 변경

IDE에서 제공해주는 equals 자동생성 기능을 도움 받아 안정적인 방식으로 변경
=> 내부적으로 LottoTicket의 List<LottNumber> 타입 간 equals됨
==> ArrayList의 equals 메서드 사용

popotec committed:
refactor(LottoMatchRank) : 메서드 리팩토링

- isFind를 인스턴스 메서드로 변경(메서드 명 변경 -> isMatchRank)
=> static 메서드로 사용하여 LottoMatchRank를 메서드 인자로 넘겨줄 필요 없이 인스턴스 메서드로 변경

- public 접근제어자로 변경하여 해당 LottoMatchRank에 메세지를 보내 비교할 수 있도록 함

popotec committed:
chore: 테스트 메서드명과 displayname 분리

popotec committed:
chore: 미사용 메서드 테스트와 같이 삭제

popotec committed:
test: 테스트시 해당 테스트 도메인 관심사 밖 데이터는 정적 팩토리 메서드를 통해 주입하도록 변경

popotec committed:
chore: 미사용 상수 제거

popotec committed:
refactor: LottoNumbersMaker에서 숫자를 추출하는 방법을 ArrangeManipulator 전략으로 위임

minsungryu commented on src/main/java/lottoservice/lottonumber/ArrangeManipulator.java:
```diff
@@ -5,4 +5,6 @@
 
 	<T> void shuffleElements(List<T> elements);
+
+	<T> List<T> getRandomElements(List<T> elements, int sizeOfPick);
```
사실 제가 의도했던건 인터페이스 행위가 추가되는 것이 아닌 인터페이스에 대한 재설계였습니다.
이 인터페이스는 이름에 구체적인 방법이 드러나서(shuffle/random)이 들어가 있어서
요구사항이 변경되거나 추가되었을 때 확장하기 어려워보이는 문제도 있습니다.

이렇게 하면 운영 코드에서는 난수를 반환하고 테스트 코드에서는 내가 테스트하고 싶은 숫자를 만들어서 반환하도록 만들 수 있지 않을까요? 예를 들면 하나의 숫자를 반환하는 인터페이스가 있다고 가정해볼게요.
```java
public interface NumberSupplier {
    int getAsInt();
}

public class RandomNumberSupplier implements NumberSupplier {

    @Override
    public int getAsInt() {
        // ... 난수를 생성하여 반환
    }
}

// 새로운 구현이 필요한 경우 추가
```

리스트를 반환하고 싶은 경우, 정렬된 숫자를 반환하고 싶은 경우에도 이런 방법을 응용해서 만들 수 있을 것 같습니다.

minsungryu commented on src/test/java/lottoservice/matcher/LottoPrizeAnswerTest.java:
```diff
@@ -5,41 +5,22 @@
-			Arrays.asList(1, 22, 3, 7, 30, 45));
-		WinningNumbers winningNumbers = new WinningNumbers(lottoNumbers);
+	public void winningNumbers_constructor_duplicate_exception() {
+		WinningNumbers winningNumbers = new WinningNumbers(TestLottoDataFactory.getLottoNumbers(1, 22, 3, 7, 30, 45));
```
제가 로또 번호들을 다 보지는 못했지만 현재 규칙이 크게 없는 것 같은데요
이런 경우는 `WinningNumbers`가 숫자의 배열을 받는 정적 팩토리 메서드를 구현한다면 `TestLottoDataFactory`가 없어도 되지 않을까요?

만약 테스트 픽스처로 관리하려면 재사용 가능한 의미있는 데이터를 만들어주시면 더 좋을 것 같습니다.
(1등 당첨번호, 1등 당첨 티켓, 2등 당첨 티켓 등)

minsungryu commented on src/test/java/lottoservice/lottonumber/LottoNumbersMakerTest.java:
```diff
@@ -16,33 +17,32 @@
 	@Test
-	public void makeLottoNumbers_1에서45사이_6개의_숫자생성() {
+	public void makeLottoNumbers_shuffle_constructor() {
 		List<LottoNumber> lottoNumbers = lottoNumbersMaker.makeLottoNumbers();
```
앞서 언급했던 인터페이스 관련 설계를 개선하면 `LottoNumbersMaker`를 테스트에 도움이 될 것 같습니다.

현재는 `new LottoArrangeManipulator()`가 고정적으로 주입되지만
각 테스트 케이스마다 `LottoNumbersMaker`가 다른 행위를 할 수 있도록 구현체를 주입해준다면
개발자가 코드의 흐름을 제어하기 쉬워지고 이를 통해 테스트도 수월해지지 않을까 생각이 듭니다.

minsungryu reviewed (approved):
지희님 로또 4단계 구현하시느라 고생많으셨습니다.
나날이 코드가 개선되는 것 같습니다 👍 
바로 merge 하겠습니다.
다음 단계도 힘내서 진행해주세요!

minsungryu commented:
앞으로 과정이 많이 남았고 체력 관리도 해야하기 때문에
제가 남긴 피드백을 구현하기 위해 너무 많은 시간을 쏟으실 필요는 없습니다.

생각해볼 거리를 드린 것이니 한 번 쓱 훑어만 보셔도 되구요~
다음 미션도 힘내서 진행해주시면 좋을 것 같아요!
